snapcraft/remote/git.py

```diff
@@ -181,23 +184,70 @@ def push_url(
             "Pushing %r to remote %r with refspec %r.", ref, stripped_url, refspec
         )
 
-        # Create a list of tags to push
+        # temporarily call git directly due to libgit2 bug that unable to push
+        # large repos using https. See https://github.com/libgit2/libgit2/issues/6385
+        # and https://github.com/snapcore/snapcraft/issues/4478
+        cmd: list[str] = ["git", "push", remote_url, refspec, "--progress"]
         if push_tags:
-            tag_refs = [
-                t.name
-                for t in self._repo.references.iterator(pygit2.GIT_REFERENCES_TAGS)
-            ]
-        else:
-            tag_refs = []
+            cmd.append("--tags")
+
+        git_proc: Optional[subprocess.Popen] = None
         try:
-            self._repo.remotes.create_anonymous(remote_url).push([refspec] + tag_refs)
-        except pygit2.GitError as error:
+            with subprocess.Popen(
+                cmd,
+                cwd=str(self.path),
+                bufsize=1,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                universal_newlines=True,
+            ) as git_proc:
+                # do not block on reading from the pipes
+                # (has no effect on Windows until Python 3.12, so the readline() method is
+                # blocking on Windows but git will still proceed)
+                if git_proc.stdout:
+                    os.set_blocking(git_proc.stdout.fileno(), False)
+                if git_proc.stderr:
+                    os.set_blocking(git_proc.stderr.fileno(), False)
+
+                git_stdout: str
+                git_stderr: str
+
+                while git_proc.poll() is None:
+                    if git_proc.stdout:
+                        while git_stdout := git_proc.stdout.readline():
+                            logger.info(git_stdout.rstrip())
+                    if git_proc.stderr:
+                        while git_stderr := git_proc.stderr.readline():
+                            logger.error(git_stderr.rstrip())
+                    # avoid too much looping, but not too slow to display progress
+                    time.sleep(0.01)
```

Comment:
re: @mr-cal's question about `craft-cli`, I think part of the trouble here is that `git push` by default outputs a lot of stuff to `stderr`. Maybe we want to either use `--porcelain` or not treat `stdout` and `stderr` differently (since git doesn't use `stderr` for error output)?